SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/SpatialRPCService.cpp

```diff
@@ -0,0 +1,484 @@
+			}
+			return EPushRPCResult::NoRingBufferAuthority;
+		}
+
+		EndpointObject = Schema_GetComponentUpdateFields(RPCStore.GetOrCreateComponentUpdate(EntityComponent));
+
+		if (Type == ERPCType::NetMulticast)
+		{
+			// Assume all multicast RPCs are auto-acked.
+			LastAckedRPCId = RPCStore.LastSentRPCIds.FindRef(EntityType);
+		}
+		else
+		{
+			// We shouldn't have authority over the component that has the acks.
+			if (AuthSubView->HasAuthority(EntityId, RPCRingBufferUtils::GetAckComponentId(Type)))
+			{
+				return EPushRPCResult::HasAckAuthority;
+			}
+
+			LastAckedRPCId = ClientServerRPCs.GetAckFromView(EntityId, Type);
+		}
+	}
+	else
+	{
+		if (bCreatedEntity)
+		{
+			return EPushRPCResult::EntityBeingCreated;
+		}
+		// If the entity isn't in the view, we assume this RPC was called before
+		// CreateEntityRequest, so we put it into a component data object.
+		EndpointObject = Schema_GetComponentDataFields(RPCStore.GetOrCreateComponentData(EntityComponent));
+
+		LastAckedRPCId = 0;
+	}
+
+	const uint64 NewRPCId = RPCStore.LastSentRPCIds.FindRef(EntityType) + 1;
+
+	// Check capacity.
+	if (LastAckedRPCId + RPCRingBufferUtils::GetRingBufferSize(Type) >= NewRPCId)
+	{
+		RPCRingBufferUtils::WriteRPCToSchema(EndpointObject, Type, NewRPCId, Payload);
+
+#if TRACE_LIB_ACTIVE
+		if (SpatialLatencyTracer != nullptr && Payload.Trace != InvalidTraceKey)
+		{
+			if (PendingTraces.Find(EntityComponent) == nullptr)
+			{
+				PendingTraces.Add(EntityComponent, Payload.Trace);
+			}
+			else
+			{
+				SpatialLatencyTracer->WriteAndEndTrace(Payload.Trace,
+													   TEXT("Multiple rpc updates in single update, ending further stack tracing"), true);
+			}
+		}
+#endif
+
+		RPCStore.LastSentRPCIds.Add(EntityType, NewRPCId);
+	}
+	else
+	{
+		// Overflowed
+		if (RPCRingBufferUtils::ShouldQueueOverflowed(Type))
+		{
+			return EPushRPCResult::QueueOverflowed;
+		}
+		else
+		{
+			return EPushRPCResult::DropOverflowed;
+		}
+	}
+
+	return EPushRPCResult::Success;
+}
+
+FRPCErrorInfo SpatialRPCService::ApplyRPC(const FPendingRPCParams& Params)
+{
+	TWeakObjectPtr<UObject> TargetObjectWeakPtr = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(Params.ObjectRef);
+	if (!TargetObjectWeakPtr.IsValid())
+	{
+		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject, ERPCQueueProcessResult::StopProcessing };
+	}
+
+	UObject* TargetObject = TargetObjectWeakPtr.Get();
+	const FClassInfo& ClassInfo = NetDriver->ClassInfoManager->GetOrCreateClassInfoByObject(TargetObjectWeakPtr.Get());
+	UFunction* Function = ClassInfo.RPCs[Params.Payload.Index];
+	if (Function == nullptr)
+	{
+		return FRPCErrorInfo{ TargetObject, nullptr, ERPCResult::MissingFunctionInfo, ERPCQueueProcessResult::ContinueProcessing };
+	}
+
+	return ApplyRPCInternal(TargetObject, Function, Params);
+}
+
+FRPCErrorInfo SpatialRPCService::ApplyRPCInternal(UObject* TargetObject, UFunction* Function, const FPendingRPCParams& PendingRPCParams)
```

Comment:
this can just be one method, right? 
though maybe it's fine. I already moved most of the content in the ApplyRPC and the ApplyRPCInternals functions into a separate `RPCUtils` class so that we can both use it and don't duplicate it.

Reply:
It can. I was about to do just that then saw quite how long the apply RPC method was, so left it as it was. Happy to merge them if you think it's a good idea, or do you think it makes more sense to merge with your changes later?